content/blog/2019-10-23-loop-unswitching/index.md

```diff
@@ -0,0 +1,252 @@
+* Bypass:  Both bypass blocks also delegate the program flow to the end of program block, essentially exiting the loop.
+* End of program:  This block is identical to the ```End of Program``` block in the original CFG.
+
+### Implementation
+
+Fine-grain details on the implementation in Python can be found [here.](https://github.com/sameerlal/bril/blob/master/bril-txt/unroll_opt.py)
+
+Implementing unswitching requires favoring generality over specificity.  We begin by loading a Bril file and storing the control flow graph by keeping track of edges.  Next, we run a dominator analysis so we can achieve, in constant time, a dictionary of blocks that dominate a particular block.  We run the loop finding algorithm as mentioned before, and then verify that it is switchable by examining conditional variables and checking to see if that variable has been used before.  We mark these blocks to be reordered and then pass the CFG into the unswitching algorithm.
+
+In my implementation, I stored the program as a dictionary mapping from block name to contents.  The unswitching algorithm takes in this mapping and produces a new mapping that is eventually converted to a Bril program.
+
+Since there are multiple duplicated blocks, we first create two hashes, one for each block, that act as suffixes for duplicated blocks.  For instance if ``` hash(if) = 083``` and ```hash(else) = 061``` then the two bypass blocks are named:  ``` bypass083``` and ```bypass061```.  This eliminates the possibility of branching to the incorrect branch, since all block names are guaranteed to be unique.  Now we are ready to create the blocks and modify branches for reordering.
+
+We first create the ```conditional t``` block, whose name has already been extracted.  
+
+Next, we create the loop logic blocks.  We copy the contents of the original program's loop logic, duplicate the block, and append the appropriate hashing suffix to its label.  The ```conditional t``` block that we just created will flow to both of these blocks, so we will need to modify the names of the branches there.
+
+```
+    if_for_loop_logic = loop_logic  + hash(if)
+    contents[if_for_loop_logic] = contents[loop_logic]
+
+    else_for_loop_logic = loop_logic  + hash(else)
+    contents[else_for_loop_logic] = contents[loop_logic]
+```
+   
+    
+Now we are ready to create the loop body blocks.  Here, the contents should include instructions that dominates the ```conditional t``` block and is also dominated by the big loop block.  Here, we use set intersections in the domination dictionary.
+    
+Now, as an additional optimization, we reorder these blocks according to the original CFG.  I will delegate the details of this for those interested to the linked code above.
```

Comment:
Aha! This answers my question from above. But I'm actually still confused about one thing: you emphasize below that the correct order needs to be preserved for correctness, but here you say that the block order is "an additional optimization." These seem to be in conflict—would the program still work correctly if you didn't carefully do the ordering? If not, then I don' think I'd call it an optimization…